File: scripts/debug-compare-joint-dynamics.py
import json
import sys
import xml.etree.ElementTree as ET
from copy import deepcopy
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
PY_BACKEND = ROOT / "server" / "python_backend"
if str(PY_BACKEND) not in sys.path:
    sys.path.insert(0, str(PY_BACKEND))

from mjwarp_runner import evaluate_controller_candidates_batch
from planner import default_tuning_trajectory
from runtime_config import load_runtime_config


def apply_joint_dynamics(raw_xml: str, joint_dynamics: dict[str, dict[str, float]]) -> str:
    root = ET.fromstring(raw_xml)
    for joint in root.iter("joint"):
        name = str(joint.get("name", "")).strip()
        if name not in joint_dynamics:
            continue
        for attr in ("damping", "frictionloss", "armature"):
            if attr in joint_dynamics[name]:
                joint.set(attr, str(float(joint_dynamics[name][attr])))
            elif attr in joint.attrib:
                joint.attrib.pop(attr, None)
    return ET.tostring(root, encoding="unicode")


def build_candidate_spec(base: list[dict]) -> dict:
    return {
        "names": [item["name"] for item in base],
        "base_gains": deepcopy(base),
        "kp": [[float(item["kp"]) for item in base]],
        "ki": [[float(item["ki"]) for item in base]],
        "kd": [[float(item["kd"]) for item in base]],
        "force_limits": [[abs(float(item["forcerange"][1])) for item in base]],
    }


def simplify(metrics: dict) -> dict:
    output = {}
    for key, value in metrics.items():
        output[key] = bool(value[0]) if str(value.dtype) == "bool" else float(value[0])
    return output


def main() -> None:
    config = load_runtime_config()
    raw_xml = (ROOT / "public" / "models" / "er15-1400.mjcf.xml").read_text("utf8")
    baseline = config["servo"]["baseline"]
    candidate_spec = build_candidate_spec(baseline)
    control_mode = {
        "ff_mode": config["backend_control"]["ff_mode"],
        "computed_torque": config["backend_control"]["computed_torque"],
        "ideal_actuation": config["backend_control"]["ideal_actuation"],
    }
    trajectory = default_tuning_trajectory()

    current_dynamics = deepcopy(config["physics"]["joint_dynamics"])
    industrial_dynamics = {
        "joint_1": {"damping": 8.0, "frictionloss": 3.2, "armature": 0.08},
        "joint_2": {"damping": 9.0, "frictionloss": 3.6, "armature": 0.08},
        "joint_3": {"damping": 8.5, "frictionloss": 3.0, "armature": 0.07},
        "joint_4": {"damping": 2.8, "frictionloss": 0.9, "armature": 0.02},
        "joint_5": {"damping": 2.0, "frictionloss": 0.5, "armature": 0.015},
        "joint_6": {"damping": 0.8, "frictionloss": 0.18, "armature": 0.006},
    }

    scenarios = {
        "no_dynamics": {},
        "current_dynamics": current_dynamics,
        "industrial_dynamics": industrial_dynamics,
    }

    results = {}
    for name, dynamics in scenarios.items():
        scene_xml = apply_joint_dynamics(raw_xml, dynamics)
        metrics = evaluate_controller_candidates_batch(
            scene_xml,
            trajectory,
            candidate_spec,
            control_mode,
            profile="full",
            return_arrays=True,
        )
        results[name] = simplify(metrics)

    print(
        json.dumps(
            {
                "control_mode": control_mode,
                "scenarios": results,
                "industrial_dynamics": industrial_dynamics,
            },
            ensure_ascii=False,
            indent=2,
        )
    )


if __name__ == "__main__":
    main()

File: scripts/debug-compare-joint-dynamics.mjs
import fs from "node:fs";
import path from "node:path";
import { spawn } from "node:child_process";

const projectRoot = "D:\\AI\\loongenv";
const pythonExe = path.join(projectRoot, ".venv-backend", "Scripts", "python.exe");
const entryPy = path.join(projectRoot, "server", "python_backend", "entry.py");
const runtimeConfigPath = path.join(projectRoot, "config", "loongenv-runtime.json");
const modelXmlPath = path.join(projectRoot, "public", "models", "er15-1400.mjcf.xml");

const originalConfigText = fs.readFileSync(runtimeConfigPath, "utf8");
const originalConfig = JSON.parse(originalConfigText);
const modelXml = fs.readFileSync(modelXmlPath, "utf8");

const industrialDynamics = {
  joint_1: { damping: 8.0, frictionloss: 3.2, armature: 0.08 },
  joint_2: { damping: 9.0, frictionloss: 3.6, armature: 0.08 },
  joint_3: { damping: 8.5, frictionloss: 3.0, armature: 0.07 },
  joint_4: { damping: 2.8, frictionloss: 0.9, armature: 0.02 },
  joint_5: { damping: 2.0, frictionloss: 0.5, armature: 0.015 },
  joint_6: { damping: 0.8, frictionloss: 0.18, armature: 0.006 },
};

const scenarios = [
  { name: "no_dynamics", joint_dynamics: {} },
  { name: "current_dynamics", joint_dynamics: originalConfig.physics?.joint_dynamics ?? {} },
  { name: "industrial_dynamics", joint_dynamics: industrialDynamics },
];

function runTune(payload) {
  return new Promise((resolve, reject) => {
    const stdout = [];
    const stderr = [];
    const child = spawn(pythonExe, [entryPy], {
      cwd: projectRoot,
      env: {
        ...process.env,
        PYTHONIOENCODING: "utf-8",
        PYTHONUTF8: "1",
        LOONGENV_BACKEND_SITE_PACKAGES: path.join(projectRoot, ".venv-backend", "Lib", "site-packages"),
      },
      stdio: ["pipe", "pipe", "pipe"],
    });

    child.stdout.on("data", (chunk) => stdout.push(String(chunk)));
    child.stderr.on("data", (chunk) => stderr.push(String(chunk)));
    child.on("exit", (code) => {
      if (code !== 0) {
        reject(new Error(`python exit ${code}\n${stderr.join("")}\n${stdout.join("")}`));
        return;
      }
      const text = stdout.join("").trim();
      try {
        resolve(JSON.parse(text));
      } catch (error) {
        reject(new Error(`failed to parse stdout as JSON\n${text}\n${String(error)}`));
      }
    });

    child.stdin.write(JSON.stringify(payload));
    child.stdin.end();
  });
}

async function main() {
  const results = {};
  try {
    for (const scenario of scenarios) {
      const patchedConfig = JSON.parse(originalConfigText);
      patchedConfig.physics = patchedConfig.physics ?? {};
      patchedConfig.physics.joint_dynamics = scenario.joint_dynamics;
      fs.writeFileSync(runtimeConfigPath, `${JSON.stringify(patchedConfig, null, 2)}\n`, "utf8");

      const payload = {
        job_id: `debug-compare-${scenario.name}`,
        task_type: "tune_controller",
        scene_prompt: "debug compare dynamics",
        task_prompt: scenario.name,
        objective_prompt: "compare pure robot tuning dynamics",
        scene_xml: modelXml,
        objective_weights: { precision: 28, stability: 28, vibration: 22, cycle: 14, energy: 8 },
        controller_gains: originalConfig.servo.baseline,
        ff_mode: originalConfig.backend_control.ff_mode,
        computed_torque: originalConfig.backend_control.computed_torque,
        ideal_actuation: originalConfig.backend_control.ideal_actuation,
        jobs: 1,
        rounds: 1,
        trials_per_round: 1,
        seed: 1234,
      };

      const result = await runTune(payload);
      results[scenario.name] = {
        joint_dynamics: scenario.joint_dynamics,
        metrics: result?.result?.metrics ?? result?.metrics ?? null,
      };
    }
  } finally {
    fs.writeFileSync(runtimeConfigPath, originalConfigText, "utf8");
  }

  console.log(
    JSON.stringify(
      {
        control_mode: originalConfig.backend_control,
        results,
      },
      null,
      2,
    ),
  );
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
